[Woo POS][Local catalog] Set up GRDB database with initial schema

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR adds GRDB along with the initial schema for POS products and variations in the catalog.

The tables aren't used anywhere, except in the tests. I think we'll still be treating them as a Work in Progress!

I've not done anything with indexes yet – I want to fully understand what we need before tackling that.

### Questions
1. I've deliberately left out fields related to stock management and sales, because they're not used in POS at the moment. Should we just include them anyway, since they're in the Networking models? If not, I think we should remove them from the Networking models for now.
2. I've removed all references to POS from the table design. That's to keep this open to future use by the main app. Do you think that's the right approach? Or would it be better to "namespace" the tables that way.
3. What do you think about the migration definition approach? I'm trying to strike a balance between "don't change this once it's shipped" and encapsulating everything in one place. We could go further with building a migration registration system, but I don't think we need more complexity yet.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Run `GRDBManagerTests`
Check CI passes

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

I've not tested this in the app yet – it needs a bit more scaffolding, e.g. `Storage.POSProduct` models, or some other name that doesn't include POS and can be converted to the existing `POSProduct` in Yosemite. This is similar to our existing Core Data approach... but it's potentially one more abstraction step than we really need – we could just make the existing POSProduct a `PersistableRecord` in a Yosemite extension...

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
Add GRDB dependency to storage

joshheald committed:
Add initial database setup

joshheald committed:
Update posProduct definition to only fields we use

joshheald committed:
Move migrations to separate files

joshheald committed:
Add ProductAttributeTable to initial migration

joshheald committed:
Add ProductImageTable to initial migration

joshheald committed:
Fix foreign key definitions

joshheald committed:
Add Variation and variation attribute tables

joshheald committed:
Add ProductVariationImage table

joshheald committed:
Remove `pos` prefix from table names

joshheald committed:
Tests for initial GRDB migration/schema creation

joshheald committed:
Add note about migration editability

dangermattic commented:
<!--
  0 Errors
  1 Warning: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald commented on Modules/Sources/Storage/GRDB/GRDBManager.swift:
```diff
@@ -0,0 +1,44 @@
+
+public final class GRDBManager {
+
+    let databaseQueue: DatabaseQueue
```
I chose DatabaseQueue instead of DatabasePool [based on the advice here](https://github.com/groue/GRDB.swift/?tab=readme-ov-file#database-connections) – a pool might end up being better for our needs, but let's see how we get on.

joshheald commented on Modules/Sources/Storage/GRDB/GRDBManager.swift:
```diff
@@ -0,0 +1,44 @@
+        try migrateIfNeeded()
+    }
+
+    init() throws {
```
Internal for use in tests

joshheald commented on Modules/Sources/Storage/GRDB/Migrations/V001InitialSchema.swift:
```diff
@@ -0,0 +1,134 @@
+            productAttributeTable.column("variation", .boolean).notNull()
+            productAttributeTable.column("options", .jsonText).notNull()
+
+            productAttributeTable.foreignKey(["siteID", "productID"], references: "product")
```
I explored using `belongsTo` instead of `foreignKey` here. It's a bit nicer, because it automatically makes the required foreign key fields on the table... but it's a bit odd with composite keys.

I think it would create `productProductID`, `productSiteID` columns in this case.

We could make a top level `Site` entity and then `Products`/`Variations` could all belong to a single site. That would let us use the `belongsTo` approach. WDYT?

jaclync replied:
> We could make a top level `Site` entity and then `Products`/`Variations` could all belong to a single site. That would let us use the `belongsTo` approach. WDYT?

Yea that makes sense to me. We can create a Site entity with just the site ID for now, and create the site before upserting the catalog.

From reading about the [`belongsTo` doc](https://swiftpackageindex.com/groue/GRDB.swift/master/documentation/grdb/belongstoassociation), it looks like it will automatically create a `belongingEntityId` foreign key column that points to the entity ID (is that where the composite primary key becomes tricky?). With `belongsTo`, we can specify the `onDelete` rule as well. Like for `product.images`, it makes sense to set it to cascade like in the existing Core Data schema.

joshheald replied:
OK, I'll give that a go. [We can set cascade deletes with `foreignKey` as well](https://swiftpackageindex.com/groue/grdb.swift/v7.6.1/documentation/grdb/columndefinition/references(_:column:ondelete:onupdate:deferred:)). `belongsTo` is a nice shorthand though.

joshheald committed:
Update Package.resolved for Storage with GRDB

jaclync commented on Modules/Sources/Storage/GRDB/GRDBManager.swift:
```diff
@@ -0,0 +1,44 @@
+        let databaseURL = URL(fileURLWithPath: databasePath)
+        let directoryURL = databaseURL.deletingLastPathComponent()
+
+        try FileManager.default.createDirectory(at: directoryURL, withIntermediateDirectories: true, attributes: nil)
```
nit: how about a test case to ensure the directory is created if it doesn't exist before?

joshheald replied:
Hmm. I kind of wanted to avoid actually messing with the filesystem in tests. Given that this is a system function, it doesn't seem like the testing is particularly worthwhile?

jaclync commented on Modules/Sources/Storage/GRDB/GRDBManager.swift:
```diff
@@ -0,0 +1,44 @@
+    func migrateIfNeeded() throws {
+        var migrator = DatabaseMigrator()
+
+        #if DEBUG
+        // Speed up development by dropping the database when migrations change
+        migrator.eraseDatabaseOnSchemaChange = true
+        #endif
```
Do we want to test migrations during development regardless of the base schema version in the app, so that we can catch any issues with migrations?
If this isn't set, does the app just crash like when launching a branch with an older schema version, and recover the next launch? I think this is fine, unless you think this has configuration has significant benefits.

joshheald replied:
I think we leave this in for initial dev, but remove it when we turn the flag on.

I want us to start with a clean slate. Let's avoid shipping the first version to customers with a bunch of migrations already, with this we can put everything in V001 and then add new ones after we ship to merchants.

jaclync commented on Modules/Sources/Storage/GRDB/Migrations/V001InitialSchema.swift:
```diff
@@ -0,0 +1,134 @@
+        try db.create(table: "product") { productTable in
+            productTable.primaryKey(["siteID", "productID"])
+
+            productTable.column("siteID", .integer).notNull()
```
just double checking, the `integer` type maps to Int64?

joshheald replied:
Essentially yes.

[It maps to `INTEGER` in SQLite](https://sqlite.org/datatype3.html). That uses 0-8 bytes, depending on the number you actually put in there. When it's read, it's always returned as an 8-byte integer, i.e. Int64.

jaclync commented on Modules/Sources/Storage/GRDB/Migrations/V001InitialSchema.swift:
```diff
@@ -0,0 +1,134 @@
+            productTable.column("name", .text).notNull()
+            productTable.column("productTypeKey", .text).notNull()
+
+            productTable.column("fullDescription", .text)
```
nit: I think we can use `description` now, which was not allowed in Core Data since it's a reserved property.

joshheald replied:
We probably can, but it's easier if they map well to the Swift types as well. `description` is the main requirement of [`CustomStringConvertable`](https://developer.apple.com/documentation/swift/customstringconvertible), which is meant for describing the Swift type. A product description isn't likely what we want for that – better would be something like `{name} - {productID}` or similar.

jaclync commented on Modules/Tests/StorageTests/GRDB/GRDBManagerTests.swift:
```diff
@@ -0,0 +1,340 @@
+    // MARK: - Migration Tests
+    struct MigrationTests {
+        @Test("Migration runs successfully")
+        func migrationRunsSuccessfully() throws {
```
nit: do we want to keep the test case names snake case to be consistent with the code base? Might not need the test annotation if the test case name is expressive enough.

jaclync commented on Modules/Tests/StorageTests/GRDB/GRDBManagerTests.swift:
```diff
@@ -0,0 +1,340 @@
+
+            // When
+            try manager.databaseQueue.write { db in
+                var record = TestProduct(
```
nit: could it be `let`?

jaclync commented on Modules/Tests/StorageTests/GRDB/GRDBManagerTests.swift:
```diff
@@ -0,0 +1,340 @@
+    }
+}
+
+// MARK: - Test Models
```
Will these test models be replaced by the production ones later, or do we intend to keep these test ones in unit tests? If we will eventually use the production models here, maybe these models can be made official?

Also, maybe these CRUD tests could be in a separate test class/file.

When we have more than one schema version, is it possible to unit test migration between versions?

joshheald replied:
> Will these test models be replaced by the production ones later, or do we intend to keep these test ones in unit tests? If we will eventually use the production models here, maybe these models can be made official?

I think we can keep the test models in these tests, even when we have the production models. It helps the tests stay more related to "does the database schema work" than "do our production models map to the database". We should test that case as well, but it's a separate thing.

joshheald replied:
> Also, maybe these CRUD tests could be in a separate test class/file.

Can do... but do you not think the separate suite here is enough? I quite like grouping them in with the schema tests – we check that it's defined properly, and then that it works...

joshheald replied:
> When we have more than one schema version, is it possible to unit test migration between versions?

I've not tried it, but it will be. 

I think the best way to do that is to to make a migrator in tests. We'd give it `n-1` migrations in the setup, migrate to that point, and then in the `When`, we'd migrate to the `n` migration that we're testing.

jaclync replied:
> > Also, maybe these CRUD tests could be in a separate test class/file.
> 
> Can do... but do you not think the separate suite here is enough? I quite like grouping them in with the schema tests – we check that it's defined properly, and then that it works...

The main benefit I see with a separate suite/file is being shorter and potentially easier to read and find for each entity. The test models can grow quite long as we support more entities.

joshheald replied:
OK. I'll look to do that in a future PR. I'm don't think we'll keep extending those CRUD tests with every single thing we add to the db – probably better to test each migration for that.

jaclync commented on Modules/Sources/Storage/GRDB/Migrations/V001InitialSchema.swift:
```diff
@@ -0,0 +1,134 @@
+            productAttributeTable.column("position", .integer).notNull()
+            productAttributeTable.column("visible", .boolean).notNull()
+            productAttributeTable.column("variation", .boolean).notNull()
+            productAttributeTable.column("options", .jsonText).notNull()
```
nice!

jaclync commented on Modules/Tests/StorageTests/GRDB/GRDBManagerTests.swift:
```diff
@@ -0,0 +1,340 @@
+        @Test("Migration runs successfully")
+        func migrationRunsSuccessfully() throws {
+            // Should not throw – initialisation performs migrations
+            let manager = try GRDBManager()
```
nit: there are a few build warnings in the test class, like this value is not used, and var -> let suggestions.

joshheald replied:
TY, I'll sort those

jaclync reviewed (approved):
Thanks for making the first GRDB PR, and in an incremental way! LGTM :shipit: 

> 1. I've deliberately left out fields related to stock management and sales, because they're not used in POS at the moment. Should we just include them anyway, since they're in the Networking models? If not, I think we should remove them from the Networking models for now.

Yes, I'm a fan of removing unused code whenever possible. Adding properties should be an incremental migration. Not sure why Periphery didn't catch the unused sales/inventory properties.

> 2. I've removed all references to POS from the table design. That's to keep this open to future use by the main app. Do you think that's the right approach? Or would it be better to "namespace" the tables that way.

Good point, yea I think this makes sense. It's hard to know the requirements in the future, it's pretty likely product management requires more fields than POS but we can always name space store management entities later on if needed. This also prepares us better if we are moving POS to a separate app.

> 3. What do you think about the migration definition approach? I'm trying to strike a balance between "don't change this once it's shipped" and encapsulating everything in one place. We could go further with building a migration registration system, but I don't think we need more complexity yet.

Looks good to start simple. As long as we know if it's possible to unit test migration between 2 versions (asked in a comment), we can improve the migration setup better later. Let's keep one schema version for the i1 launch.

joshheald committed:
Rename and fix warnings in tests

joshheald commented:
@jaclync thanks for the review. I'll work on a few simpler things inline, and bear others in mind for future PRs.

> Good point, yea I think this makes sense. It's hard to know the requirements in the future, it's pretty likely product management requires more fields than POS but we can always name space store management entities later on if needed. This also prepares us better if we are moving POS to a separate app.

Yes to both.

I think product management could probably work with a lot less fields than it has right now. It may not need them all to be stored locally, even if it needs some for occasional updates. 

For example, perhaps we would only store what's needed to show the Product Details screen. Subscreens dealing with other details have to fetch from the remote to do their work.

But with this approach, we have options to namespace later – e.g. we could have `product` and `fullProduct`, named for what they have, not where they're used.

joshheald commented:
> Yes, I'm a fan of removing unused code whenever possible. Adding properties should be an incremental migration. Not sure why Periphery didn't catch the unused sales/inventory properties.

I think it's probably because they're "used" in generated code, or mapping code, or something...

I'll remove the fields in a separate PR.

joshheald committed:
Add site entity to database

joshheald committed:
Simplify readable test names

joshheald committed:
Temporary Periphery ignore

joshheald commented:
@jaclync I've added a site entity and updated the keys and relationships here.

In doing this, I realised that we need to handle both local and global product attributes – I'm going to try making separate tables for these with a view to put them together in the application layer. That'll be in a different PR. 

For now, we only handle local attributes. For the purposes of POS, it might be enough to treat global attributes as local, though it'd mean a lot of duplication in the database. It wouldn't work well for future product editing, and I'm not sure we should start in this way, but it is one option I'll consider.

jaclync commented:
> In doing this, I realised that we need to handle both local and global product attributes – I'm going to try making separate tables for these with a view to put them together in the application layer. That'll be in a different PR.

Are they not all in the equivalent Product's`attributes: [ProductAttribute]` field? I think local ones have ID 0. Or did you mean something different?

joshheald commented:
> Are they not all in the equivalent Product's`attributes: [ProductAttribute]` field? I think local ones have ID 0. Or did you mean something different?

They are, but handling them correctly in the database is tricky.

If our `productAttribute` table holds both global and local attributes, we can't:
1. Use the `id` from the network as a primary key (we can't insert a bunch of attributes with id 0 if it's the PK.)
2. Have a many:many relationship from products to global attributes.

We could get round both of those things.
1. Ignore the `id` from the network, and use an autoincremented ID as the PK.
2. Have a 1:many relationship from product to attributes.

But by doing that, we'd be making a new copy of each global attribute every time it's used, which is wasteful. We also make it more difficult to deal with attributes via store management in future.

I think that solution is potentially fine for POS, though I don't like the duplication. We don't have to reference attributes by ID at the moment.... 

But I think if we can model it correctly without huge overhead, we should try to do that – so it's on the list for next week. There are a few ways to try, but it'll involve at least two tables for attributes.

jaclync commented:
> If our `productAttribute` table holds both global and local attributes, we can't:
> 
> 1. Use the `id` from the network as a primary key (we can't insert a bunch of attributes with id 0 if it's the PK.)
> 2. Have a many:many relationship from products to global attributes.
> 
> We could get round both of those things.
> 
> 1. Ignore the `id` from the network, and use an autoincremented ID as the PK.
> 2. Have a 1:many relationship from product to attributes.
> 
> But by doing that, we'd be making a new copy of each global attribute every time it's used, which is wasteful. We also make it more difficult to deal with attributes via store management in future.
> 
> I think that solution is potentially fine for POS, though I don't like the duplication. We don't have to reference attributes by ID at the moment....
> 
> But I think if we can model it correctly without huge overhead, we should try to do that – so it's on the list for next week. There are a few ways to try, but it'll involve at least two tables for attributes.

After more thinking, I think we could:

1. Use an autoincremented ID as the PK, and a separate column like `remote_id` from the network.
2. Have a many:many relationship between products and attributes for global attributes, by adding a join table like [this GRDB example](https://github.com/groue/GRDB.swift/issues/1063#issuecomment-925735039).

@joshheald WDYT?

joshheald commented:
> After more thinking, I think we could:
> 
>     1. Use an autoincremented ID as the PK, and a separate column like `remote_id` from the network.
> 
>     2. Have a many:many relationship between products and attributes for global attributes, by adding a join table like [this GRDB example](https://github.com/groue/GRDB.swift/issues/1063#issuecomment-925735039).


Yep, that's the other way – I should have mentioned this in my last comment. 

I prefer using the real ID as the PK for global attributes, but that would require separate tables, which isn't so good. Thinking about it more, I think you're right, this is probably a better trade off as the relationships will work the same for each type. I'll be looking at proper global support soon, and I'll try it then 😊
